File: top-100-liked/typescript/55-jump-game.ts
/**
 * 跳跃游戏
 * https://leetcode.cn/problems/jump-game/
 */

export function canJump(nums: number[]): boolean {
  const n = nums.length
  let rightMost = 0
  for (let i = 0; i < n; i++) {
    if (i <= rightMost) {
      rightMost = Math.max(rightMost, i + nums[i])
      if (rightMost >= n - 1) {
        return true
      }
    }
  }
  return false
}

const nums = [2, 3, 1, 1, 4]
console.log(canJump(nums))

File: top-100-liked/python/55-jump-game.py
"""
55. 跳跃游戏
https://leetcode.cn/problems/jump-game
"""
from typing import List


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n, mostJumped = len(nums), 0
        for i in range(n):
            if i <= mostJumped:
                mostJumped = max(mostJumped, i + nums[i])
                if mostJumped >= n - 1:
                    return True
        return False
